pkg/backoff/exponential.go

```diff
@@ -15,45 +15,94 @@
 // Package backoff is a wrapper of `github.com/cenkalti/backoff/v4`.
 // It is to prevent misuse of `github.com/cenkalti/backoff/v4`,
 // thus application could fall into dead loop.
-// The difference is in this package `NextBackOff` returns a MaxDuration rather than backoff.Stop(which is -1)
-// when MaxElapsedTime elapsed.
 package backoff
 
 import (
+	"context"
+	"fmt"
 	"time"
 
 	"github.com/cenkalti/backoff/v4"
 )
 
 // BackOff is a backoff policy for retrying an operation.
 type BackOff interface {
-	// NextBackOff returns the duration to wait before retrying the operation,
-	// or backoff. Return MaxDuration when MaxElapsedTime elapsed.
+	// NextBackOff returns the duration to wait before retrying the next operation.
 	NextBackOff() time.Duration
+	// Reset to initial state.
+	Reset()
+	// RetryWithContext tries the operation until it does not return error,
+	// or when the context expires, whichever happens first.
+	RetryWithContext(ctx context.Context, operation func() error) error
+}
+
+type Option struct {
+	InitialInterval time.Duration
+	MaxInterval     time.Duration
 }
 
 // ExponentialBackOff is a wrapper of backoff.ExponentialBackOff to override its NextBackOff().
 type ExponentialBackOff struct {
-	*backoff.ExponentialBackOff
+	exponentialBackOff *backoff.ExponentialBackOff
 }
 
-// NewExponentialBackOff creates an istio wrapped ExponentialBackOff.
-func NewExponentialBackOff(initFuncs ...func(off *ExponentialBackOff)) BackOff {
-	b := &ExponentialBackOff{}
-	b.ExponentialBackOff = backoff.NewExponentialBackOff()
-	for _, fn := range initFuncs {
-		fn(b)
+// Default values for ExponentialBackOff.
+const (
+	defaultInitialInterval = 500 * time.Millisecond
+	defaultMaxInterval     = 60 * time.Second
+)
+
+func DefaultOption() Option {
+	return Option{
+		InitialInterval: defaultInitialInterval,
+		MaxInterval:     defaultMaxInterval,
 	}
+}
+
+// NewExponentialBackOff creates an istio wrapped ExponentialBackOff.
+// By default, it never stops.
+func NewExponentialBackOff(o Option) BackOff {
+	b := ExponentialBackOff{}
+	b.exponentialBackOff = backoff.NewExponentialBackOff()
+	b.exponentialBackOff.InitialInterval = o.InitialInterval
+	b.exponentialBackOff.MaxInterval = o.MaxInterval
 	b.Reset()
 	return b
 }
 
-const MaxDuration = 1<<63 - 1
-
-func (b *ExponentialBackOff) NextBackOff() time.Duration {
-	duration := b.ExponentialBackOff.NextBackOff()
-	if duration == b.Stop {
-		return MaxDuration
+func (b ExponentialBackOff) NextBackOff() time.Duration {
+	duration := b.exponentialBackOff.NextBackOff()
+	if duration == b.exponentialBackOff.Stop {
+		return b.exponentialBackOff.MaxInterval
 	}
 	return duration
 }
+
+func (b ExponentialBackOff) Reset() {
+	b.exponentialBackOff.Reset()
+}
+
+// RetryWithContext tries the operation until it does not return error,
+// or when the context expires, whichever happens first.
```

Comment:
currently it also stops when hitting maxInterval I think?

Reply:
Nope, actually  no way to set maxElapsedTime now. I think this can be removed